euphoria-spark/src/main/java/cz/seznam/euphoria/spark/accumulators/SparkNativeAccumulators.java

```diff
@@ -0,0 +1,165 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package cz.seznam.euphoria.spark.accumulators;
+
+import cz.seznam.euphoria.core.client.accumulators.Accumulator;
+import cz.seznam.euphoria.core.client.accumulators.AccumulatorProvider;
+import cz.seznam.euphoria.core.client.accumulators.Counter;
+import cz.seznam.euphoria.core.client.accumulators.Histogram;
+import cz.seznam.euphoria.core.client.accumulators.Timer;
+import cz.seznam.euphoria.core.util.Settings;
+import org.apache.spark.api.java.JavaSparkContext;
+import org.apache.spark.util.AccumulatorV2;
+
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Integrates the native Spark accumulators. Accumulators will be
+ * visible in the Spark UI.
+ */
+public class SparkNativeAccumulators implements AccumulatorProvider, Serializable {
+
+  private final SparkAccumulatorHolder accumulators;
+
+  public SparkNativeAccumulators(SparkAccumulatorHolder accumulators) {
+    this.accumulators = accumulators;
+  }
+
+  @Override
+  public Counter getCounter(String name) {
+    return accumulators.getCounter(name);
+  }
+
+  @Override
+  public Histogram getHistogram(String name) {
+    return accumulators.getHistogram(name);
+  }
+
+  @Override
+  public Timer getTimer(String name) {
+    return accumulators.getTimer(name);
+  }
+
+  public static class Factory implements SparkAccumulatorFactory {
+
+    private static final Factory INSTANCE = new Factory();
+
+    // ~ since SparkNativeAccumulators is serializable class
+    // it can be directly embedded into the factory
+    private SparkNativeAccumulators provider;
+
+    private Factory() {}
+
+    public static Factory get() {
+      return INSTANCE;
+    }
+
+    @Override
+    public AccumulatorProvider create(Settings settings) {
+      if (provider == null) {
+        throw new IllegalStateException("Accumulator factory not initialized.");
+      }
+
+      return provider;
+    }
+
+    @Override
+    public void init(JavaSparkContext sparkContext) {
+      SparkAccumulatorHolder sparkAccumulatorHolder = new SparkAccumulatorHolder();
+      sparkContext.sc().register(sparkAccumulatorHolder, "accumulators");
+
+      provider = new SparkNativeAccumulators(sparkAccumulatorHolder);
+    }
+  }
+
+  /**
+   * This class contains a map of accumulators. Spark requires all accumulators
+   * to be declared before the job is executed. This holder is created before
+   * the job is launched and registered to Spark context allowing Euphoria
+   * to add accumulators on the fly.
+   */
+  private static class SparkAccumulatorHolder
```

Comment:
* looks good to me 👍 
* the only think i'm a bit worried about is the docs of `AccumulatorV2` stating ... _`OUT` should be a type that can be read atomically (e.g., Int, Long), or thread-safely (e.g., synchronized collections) because it will be read from other threads._ have you had any chance to investigate how accumulators are really accessed?